### DISCUSSION TITLE
CefSharp upgrade and getting unhandled exception Cannot pass a GCHandle across AppDomains

### DISCUSSION THREAD

**Discussion by Dileepreddyj** [cefsharp/CefSharp#4719] [Q&A]
We have been utilizing CefSharp version 71 for approximately four years and recently made the decision to upgrade to version 120. While implementing the required modifications, we encountered an unhandled exception during the initialization of Cef (Cef.Initialize), specifically stating 'Cannot pass a GCHandle across AppDomains.' Despite having come across articles in the past discussing the same 'Cannot pass a GCHandle across AppDomains' exception with CefSharp, we have not yet found a suitable solution. 

Adding the in-depth error for reference:

System.ArgumentException
  HResult=0x80070057
  Message=Cannot pass a GCHandle across AppDomains.
Parameter name: handle
  Source=mscorlib
  StackTrace:
   at System.Runtime.InteropServices.GCHandle.InternalCheckDomain(IntPtr handle)
   at System.Runtime.InteropServices.GCHandle.FromIntPtr(IntPtr value)
   at gcroot<System::Collections::Generic::IEnumerable<CefSharp::CefCustomScheme ^> ^>..P$AAU?$IEnumerable@P$AAVCefCustomScheme@CefSharp@@@Generic@Collections@System@@(gcroot<System::Collections::Generic::IEnumerable<CefSharp::CefCustomScheme ^> ^>* )
   at CefSharp.Internals.CefSharpApp.OnContextInitialized(CefSharpApp* )

**Was able to reproduce the issue in a sample app. Sample code that can be added while initializing cef to reproduce the issue.**

  try
            {
                AppDomain domain = AppDomain.CreateDomain("another domain");
                CrossAppDomainDelegate action = () => { Cef.Initialize(new CefSettings()); };
                domain.DoCallBack(action);
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }

**We would greatly appreciate assistance in comprehending and resolving this issue.**

**Comment by amaitland** [comment]
Running `CefSharp` in non default app domain is not supported. This is a limitation of `Visual C++` which `CefSharp` is built with.

No official release of `CefSharp` has ever supported being run in the non default app domain.

**Reply by Dileepreddyj** [reply]
@amaitland
Thank you for your prompt response. I have gained clarity that running CefSharp in a non-default app domain is not supported. The provided sample code served as a means to replicate the issue outside the application. What confuses me is that, despite solely updating the older version 71 DLLs and the necessary supporting CefSharp files to those of version 120 without any code-level alterations, we have encountered the gchandle issue. Can you identify any unforeseen changes that might have caused CefSharp to operate in a different app domain?

**Reply by amaitland** [reply]
Have you confirmed that `Cef.Initialzie` is called in the default app domain? The error suggest that you are using a non-default app domain.

> Can you identify any unforeseen changes that might have caused CefSharp to operate in a different app domain?

No.
